Don't start config thread until _after_ components are set up

vekterli opened the pull request:
@baldersheim please review
@hakonhall FYI

Avoids race where new config received after initial fetch but before node is fully set up gets lost.

vekterli committed:
Don't start config thread until _after_ components are set up

Avoids race where new config received after initial fetch but before
node is fully set up gets lost.

baldersheim reviewed (approved):
I also seem remember an if where handleLiveC.. was called from that could either be an assert or be removed.

vekterli commented:
That would be the check for whether the `configure()` call is the first (synchronous) subscription callback, or a subsequent asynchronous reconfig. This should still work, but now without the potential of happening before the components are set up.

vekterli commented:
storageserver tests are complaining, will check it out

vekterli committed:
Protect initial setup from reconfigs with explicit mutex

Mutex will only be attempted concurrently taken if a live reconfig
happens during node initialize(), and will not deadlock when config
is initially propagated via the ConfigFetcher.

vekterli commented:
@baldersheim please take another look.

Semantics of config subscriptions are different than I originally thought, i.e. no sync callback upon `subscribe()`. Callbacks are not invoked until `ConfigFetcher::start()` is called, after which point async callbacks may happen at any time. Redesigned solution in the pragmatic way most problems are handled in multiprogramming: add another mutex! 🎉

baldersheim commented on storage/src/vespa/storage/storageserver/storagenode.cpp:
```diff
@@ -247,11 +252,10 @@ StorageNode::updateUpgradeFlag(const vespa::config::content::UpgradingConfig& co
 void
 StorageNode::handleLiveConfigUpdate()
 {
```
I think it wise signal the lock must be held in the signature.

vekterli replied:
Agreed. The less we allow gun vs. foot interaction, the better.

baldersheim committed:
- Add the mandatory guard to th emethod signature.
- (smartptr.get() != nullptr/0)  -> (smartptr)
- smartptr.reset(0) -> smartptr.reset()
- Avoid excessive vertical bloat.

baldersheim commented:
@vekterli I added a pull request with the guard in the signature. I caught usage from ServiceLayerNode that I think needed the same guard.

baldersheim commented:
from my point of view this is now ready to go.

baldersheim committed:
Simplify with using directive.
